core/rapid_test.go

```diff
@@ -383,15 +395,209 @@ func TestProperty_MajorityHonestNodes(t *testing.T) {
 
 			// Wait until Quorum nodes finish their run loop
 			ctx, cancelFn := context.WithTimeout(context.Background(), time.Second*5)
-			if err := cluster.awaitNCompletions(ctx, int64(quorum(numNodes))); err != nil {
-				t.Fatalf(
-					fmt.Sprintf(
-						"unable to wait for nodes to complete, %v",
-						err,
-					),
+			err := cluster.awaitNCompletions(ctx, int64(quorum(numNodes)))
+			require.NoError(t, err, "unable to wait for nodes to complete")
+
+			// Shutdown the remaining nodes that might be hanging
+			cluster.forceShutdown()
+			cancelFn()
+		}
+
+		// Make sure that the inserted proposal is valid for each height
+		for _, proposalMap := range insertedProposals.proposals {
+			for _, insertedProposal := range proposalMap {
+				assert.Equal(t, message.proposal, insertedProposal)
+			}
+		}
+	})
+}
+
+// TestProperty_MajorityHonestNodes_BadProposal is a property-based test
+// that assures the cluster can reach consensus on any
+// arbitrary number of valid nodes and byzantine nodes
+func TestProperty_MajorityHonestNodes_BadProposal(t *testing.T) {
+	t.Parallel()
+
+	rapid.Check(t, func(t *rapid.T) {
+		var multicastFn func(message *proto.Message)
+
+		var (
+			correctMessage = roundMessage{
+				proposal: []byte("proposal"),
+				hash:     []byte("proposal hash"),
+				seal:     []byte("seal"),
+			}
+
+			badMessage = roundMessage{
+				proposal: []byte("bad proposal"),
+				hash:     []byte("bad proposal hash"),
+				seal:     []byte("bad seal"),
+			}
+
+			numNodes          = rapid.Uint64Range(4, 30).Draw(t, "number of cluster nodes")
+			numByzantineNodes = rapid.Uint64Range(1, maxFaulty(numNodes)).Draw(t, "number of byzantine nodes")
+			desiredHeight     = rapid.Uint64Range(1, 5).Draw(t, "minimum height to be reached")
+
+			nodes             = generateNodeAddresses(numNodes)
+			insertedProposals = newMockInsertedProposals(numNodes)
+		)
+
+		commonLoggerCallback := func(logger *mockLogger) {
+			logger.infoFn = func(s string, i ...interface{}) {
+				fmt.Println(append([]interface{}{s}, i...)...)
+				t.Logf(s, i...)
+			}
+
+			logger.errorFn = func(s string, i ...interface{}) {
+				fmt.Println(append([]interface{}{s}, i...)...)
+				t.Errorf(s, i...)
+			}
+
+			logger.debugFn = func(s string, i ...interface{}) {
+				fmt.Println(append([]interface{}{s}, i...)...)
+				t.Logf(s, i...)
+			}
+		}
+
+		// commonTransportCallback is the common method modification
+		// required for Transport, for all nodes
+		commonTransportCallback := func(transport *mockTransport) {
+			transport.multicastFn = func(message *proto.Message) {
+				multicastFn(message)
+			}
+		}
+
+		// commonBackendCallback is the common method modification required
+		// for the Backend, for all nodes
+		commonBackendCallback := func(backend *mockBackend, nodeIndex int) {
+			message := correctMessage
+			if uint64(nodeIndex) < numByzantineNodes {
+				message = badMessage
+			}
```

Comment:
(1/3 - 1) bad nodes send bad message during round

Reply:
But single round. This is possible to write without property based testing.

Reply:
do we have a task to extend property based and have a random sequence of events normal and malicious?

Reply:
Just created one: EVM-195